concrete/src/Entity/File/Version.php

```diff
@@ -657,697 +872,862 @@ public function getVersionLogComments()
+        ++$fvID;
+
+        $fv = clone $this;
+        $fv->fvID = $fvID;
+        $fv->fvIsApproved = false;
+        $fv->fvDateAdded = new DateTime();
+        if (User::isLoggedIn()) {
+            $uID = (int) (new User())->getUserID();
+            if ($uID !== 0) {
+                $fv->fvAuthorUID = $uID;
+            }
         }
 
-        $this->fvIsApproved = true;
-        $uID = (int) (new User())->getUserID();
-        if ($uID !== 0) {
-            $this->fvApproverUID = $uID;
+        $em->persist($fv);
+
+        $this->deny();
+
+        foreach ($this->getAttributes() as $value) {
+            $value = clone $value;
+            $value->setVersion($fv);
+            $em->persist($value);
         }
-        $this->save();
 
-        $fe = new \Concrete\Core\File\Event\FileVersion($this);
-        Events::dispatch('on_file_version_approve', $fe);
+        $em->flush();
 
-        $fo = $this->getFile();
-        $fo->reindex();
+        $fve = new FileVersionEvent($fv);
+        $app->make(EventDispatcherInterface::class)->dispatch('on_file_version_duplicate', $fve);
 
-        \Core::make('cache/request')->delete('file/version/approved/'.$this->getFileID());
+        return $fv;
     }
 
     /**
-     * Return the contents of a file.
-     *
-     * @return string
+     * Duplicate the underlying file and assign its new position to this instance.
      */
-    public function getFileContents()
+    public function duplicateUnderlyingFile()
     {
-        $cf = Core::make('helper/concrete/file');
-        $fsl = $this->getFile()->getFileStorageLocationObject();
-        if (is_object($fsl)) {
-            return $fsl->getFileSystemObject()->read($cf->prefix($this->fvPrefix, $this->fvFilename));
-        }
+        $app = Application::getFacadeApplication();
+        $importer = new Importer();
+        $fi = $app->make('helper/file');
+        $cf = $app->make('helper/concrete/file');
+        $filesystem = $this->getFile()->getFileStorageLocationObject()->getFileSystemObject();
+        $fileName = $this->getFileName();
+        do {
+            $prefix = $importer->generatePrefix();
+            $path = $cf->prefix($prefix, $fileName);
+        } while ($filesystem->has($path));
+        $fileContents = $this->getFileResource()->read();
+        $mimeType = Util::guessMimeType($fileName, $fileContents);
+        $filesystem->write(
+            $path,
+            $fileContents,
+            [
+                'visibility' => AdapterInterface::VISIBILITY_PUBLIC,
+                'mimetype' => $mimeType,
+            ]
+        );
+        $this->updateFile($fileName, $prefix);
     }
 
     /**
-     * Returns a url that can be used to download a file, will force the download of all file types, even if your browser can display them.
+     * Delete this version of the file.
      *
-     * @return string
+     * @param bool $deleteFilesAndThumbnails should we delete the actual file and the thumbnails?
```

Comment:
'files and thumbnails', or 'file and thumbnails'?

Reply:
`the actual file` stands for the real file stored in the file system.
`the thumbnails` stands for the (possibly existing) thumbnails associated to the actual file.
So, `file` is singular, and `thumbnails` is plural.
We could say `the actual file and thumbnails`, but `actual` is only referred to the file and not to the thumbnails, that's why I wrote `the actual file and the thumbnails`